src/limbo/acqui/ei.hpp

```diff
@@ -0,0 +1,115 @@
+//| license as circulated by CEA, CNRS and INRIA at the following URL
+//| "http://www.cecill.info".
+//|
+//| As a counterpart to the access to the source code and  rights to copy,
+//| modify and redistribute granted by the license, users are provided only
+//| with a limited warranty  and the software's author,  the holder of the
+//| economic rights,  and the successive licensors  have only  limited
+//| liability.
+//|
+//| In this respect, the user's attention is drawn to the risks associated
+//| with loading,  using,  modifying and/or developing or reproducing the
+//| software by the user in light of its specific status of free software,
+//| that may mean  that it is complicated to manipulate,  and  that  also
+//| therefore means  that it is reserved for developers  and  experienced
+//| professionals having in-depth computer knowledge. Users are therefore
+//| encouraged to load and test the software's suitability as regards their
+//| requirements in conditions enabling the security of their systems and/or
+//| data to be ensured and,  more generally, to use and operate it in the
+//| same conditions as regards security.
+//|
+//| The fact that you are presently reading this means that you have had
+//| knowledge of the CeCILL-C license and that you accept its terms.
+//|
+#ifndef LIMBO_ACQUI_EI_HPP
+#define LIMBO_ACQUI_EI_HPP
+
+#include <cmath>
+
+#include <Eigen/Core>
+
+#include <limbo/tools/macros.hpp>
+
+namespace limbo {
+    namespace defaults {
+        struct acqui_ei {
+            /// @ingroup acqui_defaults
+            BO_PARAM(double, jitter, 0.0);
+        };
+    }
+    namespace acqui {
+        /** @ingroup acqui
+        \rst
+        Classic EI (Expected Improvement). See :cite:`brochu2010tutorial`, p. 14
+
+          .. math::
+            EI(x) = (\mu(x) - f(x^+) - \xi)\Phi(Z) + \sigma(x)\phi(Z),\\\text{with } Z = \frac{\mu(x)-f(x^+) - \xi}{\sigma(x)}.
+
+        Parameters:
+          - ``double jitter`` - :math:`\xi`
+        \endrst
+        */
+        template <typename Params, typename Model>
+        class EI {
+        public:
+            EI(const Model& model, int iteration = 0) : _model(model) {}
+
+            size_t dim_in() const { return _model.dim_in(); }
+
+            size_t dim_out() const { return _model.dim_out(); }
+
+            template <typename AggregatorFunction>
+            double operator()(const Eigen::VectorXd& v, const AggregatorFunction& afun) const
+            {
+                Eigen::VectorXd mu;
+                double sigma_sq;
+                std::tie(mu, sigma_sq) = _model.query(v);
+                double sigma = std::sqrt(sigma_sq);
+
+                // If \sigma(x) = 0 or we do not have any observation yet we return 0
+                if (sigma < 1e-10 || _model.observations().size() < 1)
+                    return 0.0;
+
+                // Compute EI(x)
+                // First find the best so far observation
```

Comment:
Instead of computing every time we call the acquisition function the best observation so far, maybe we can store the information in the GP. What do you think? I cannot really think of any multi-threading issue, but having more pair of eyes looking at this it is probably a good idea.